routing/dht/provide_many.go

```diff
@@ -0,0 +1,202 @@
+}
+
+type provManyReq struct {
+	closest     map[key.Key]pqueue.PeerQueue
+	perKeyQuery map[key.Key]pqueue.PeerQueue
+	queried     *pset.PeerSet
+
+	keys    []key.Key
+	keyStrs []string
+
+	nextTarget int
+	gntLock    sync.Mutex
+}
+
+func newProvManyReq(ctx context.Context, dht *IpfsDHT, keys []key.Key) *provManyReq {
+	closest := make(map[key.Key]pqueue.PeerQueue)
+	perKeyQuery := make(map[key.Key]pqueue.PeerQueue)
+	var keyStrs []string
+
+	for _, k := range keys {
+		keyStrs = append(keyStrs, string(k))
+		dht.providers.AddProvider(ctx, k, dht.self)
+
+		closest[k] = pqueue.NewXORDistancePQ(k)
+		perKeyQuery[k] = pqueue.NewXORDistancePQ(k)
+
+		peers := dht.routingTable.NearestPeers(kb.ConvertKey(k), 20)
+		for _, p := range peers {
+			closest[k].Enqueue(p)
+			perKeyQuery[k].Enqueue(p)
+		}
+	}
+
+	return &provManyReq{
+		closest:     closest,
+		perKeyQuery: perKeyQuery,
+		keys:        keys,
+		keyStrs:     keyStrs,
+		queried:     pset.New(),
+	}
+}
+
+func (pmr *provManyReq) getNextTarget() (peer.ID, bool) {
+	pmr.gntLock.Lock()
+	defer pmr.gntLock.Unlock()
+	// iterate through entire list once, starting at last offset
+	for i := pmr.nextTarget + 1; i != pmr.nextTarget; i = (i + 1) % len(pmr.keys) {
+		k := pmr.keys[i]
+		for pmr.perKeyQuery[k].Len() > 0 {
+			p := pmr.perKeyQuery[k].Dequeue()
+			if pmr.queried.TryAdd(p) {
+				pmr.nextTarget = i
+				return p, true
+			}
+		}
+	}
+
+	return "", false
+}
+
+func (pmr *provManyReq) addCloserPeers(ps []peer.ID) {
+	pmr.gntLock.Lock()
+	defer pmr.gntLock.Unlock()
+	for _, p := range ps {
+		if pmr.queried.Contains(p) {
+			continue
+		}
+
+		for _, k := range pmr.keys {
+			pmr.perKeyQuery[k].Enqueue(p)
+			pmr.closest[k].Enqueue(p)
+		}
+	}
+}
+
+func (pmr *provManyReq) finalProvideSet() map[peer.ID][]key.Key {
+	final := make(map[peer.ID][]key.Key)
+	for k, cq := range pmr.closest {
+		for i := 0; i < KValue && cq.Len() > 0; i++ {
+			p := cq.Dequeue()
+			final[p] = append(final[p], k)
+		}
+	}
+
+	return final
+}
+
+func (dht *IpfsDHT) ProvideMany(ctx context.Context, keys []key.Key) error {
+	defer log.EventBegin(ctx, "provideMany").Done()
+
+	pmreq := newProvManyReq(ctx, dht, keys)
+
+	t := pb.Message_FIND_NODE
+	mes := &pb.Message{
+		Type: &t,
+		Keys: pmreq.keyStrs,
+	}
+
+	query := dht.newQuery("", func(ctx context.Context, p peer.ID) (*dhtQueryResult, error) {
+		resp, err := dht.sendRequest(ctx, p, mes)
+		if err != nil {
+			return nil, err
+		}
+
+		peers := pb.PBPeersToPeerInfos(resp.GetCloserPeers())
+		var pids []peer.ID
+		for _, clpeer := range peers {
+			dht.peerstore.AddAddrs(clpeer.ID, clpeer.Addrs, peer.TempAddrTTL)
+			pids = append(pids, clpeer.ID)
+		}
+
+		pmreq.addCloserPeers(pids)
+
+		result := new(dhtQueryResult)
+		next, ok := pmreq.getNextTarget()
+		if ok {
+			result.closerPeers = []peer.PeerInfo{{ID: next}}
+		}
+
+		return result, nil
+	})
+
+	dqr := dhtQueryRunner{
+		query:          query,
+		peersToQuery:   pqueue.NewChanQueue(ctx, new(peerFifo)),
```

Comment:
because query no longer has an xor target and is instead shotgunning the entire dht? this is not sensible. the whole point of kademlia is lost-- the queue sorting is what gives you a gradient. this will fall on its face when hitting large numbers of nodes.